File: supplychain-backend/indexer/indexer.py
import os, time, json
from web3 import Web3
from dotenv import load_dotenv
load_dotenv()

GANACHE = os.getenv('GANACHE_URL','http://127.0.0.1:7545')
w3 = Web3(Web3.HTTPProvider(GANACHE))
print('Connected:', w3.is_connected())

ABI_PATH = os.path.join(os.path.dirname(__file__), 'abi.json')
CONTRACT_ADDR = os.getenv('SUPPLYCHAIN_ADDRESS','')

if os.path.exists(ABI_PATH):
    abi = json.load(open(ABI_PATH))
else:
    abi = []

if not CONTRACT_ADDR:
    print('Set SUPPLYCHAIN_ADDRESS in env to listen for events.')

def poll():
    # demo: print latest block number periodically
    last = w3.eth.block_number
    while True:
        if w3.eth.block_number > last:
            print('New block', w3.eth.block_number)
            last = w3.eth.block_number
        time.sleep(2)

if __name__=='__main__':
    poll()

File: supplychain-backend/indexer/indexer.js
const { ethers } = require("ethers");
require("dotenv").config();

// 1. Kết nối với Ganache
const GANACHE_URL = process.env.GANACHE_URL || "http://127.0.0.1:7545";
const provider = new ethers.providers.JsonRpcProvider(GANACHE_URL);

// Lấy thông tin contract
const ABI = require("./abi.json");
const CONTRACT_ADDR = process.env.SUPPLYCHAIN_ADDRESS || "";

// Kiểm tra nếu thiếu địa chỉ contract
if (!CONTRACT_ADDR) {
  console.log(
    "Vui lòng đặt biến SUPPLYCHAIN_ADDRESS trong file .env để lắng nghe sự kiện."
  );
}

// Hàm poll để kiểm tra block mới
async function poll() {
  // Kiểm tra kết nối ban đầu
  try {
    const network = await provider.getNetwork();
    console.log("Đã kết nối tới mạng:", network.name);
  } catch (error) {
    console.error(
      "Không thể kết nối tới provider. Vui lòng kiểm tra lại GANACHE_URL."
    );
    return;
  }

  let lastBlock = await provider.getBlockNumber();
  console.log(`Bắt đầu theo dõi từ block: ${lastBlock}`);

  // Sử dụng setInterval để kiểm tra định kỳ mỗi 2 giây
  setInterval(async () => {
    const currentBlock = await provider.getBlockNumber();
    if (currentBlock > lastBlock) {
      console.log(`Block mới được tìm thấy: ${currentBlock}`);
      lastBlock = currentBlock;
    }
  }, 2000);
}

poll();
